File: modules/core/src/encoder/LangEncoder.ts
import {
  Encoder,
  IntEncoder,
} from '.';

import {
  DecodingError,
  EncodingError,
} from '../errors';

export class LangEncoder implements Encoder<string> {

  public encode(value: string, numBits: number): string {

    value = value.toUpperCase();
    const intEncoder: IntEncoder = new IntEncoder();

    const ASCII_START = 65;
    const firstLetter: number = value.charCodeAt(0) - ASCII_START;

    const secondLetter: number = value.charCodeAt(1) - ASCII_START;

    // check some things to throw some good errors
    if (firstLetter < 0 || firstLetter > 25 || secondLetter < 0 || secondLetter > 25) {

      throw new EncodingError(`invalid language code: ${value}`);

    }

    if (numBits % 2 === 1) {

      throw new EncodingError(`numBits must be even, ${numBits} is not valid`);

    }

    numBits = numBits/2;
    const firstLetterBString: string = intEncoder.encode(firstLetter, numBits);
    const secondLetterBString: string = intEncoder.encode(secondLetter, numBits);

    return firstLetterBString + secondLetterBString;

  }

  public decode(value: string): string {

    let retr: string;

    // is it an even number of bits? we have to divide it
    if (!(value.length % 2)) {

      const intDecoder: IntEncoder = new IntEncoder();
      const ASCII_START = 65;
      const mid: number = value.length/2;
      const firstLetter = intDecoder.decode(value.slice(0, mid)) + ASCII_START;
      const secondLetter = intDecoder.decode(value.slice(mid)) + ASCII_START;

      retr = String.fromCharCode(firstLetter) + String.fromCharCode(secondLetter);

    } else {

      throw new DecodingError('invalid bit length for language');

    }

    return retr;

  }

}

File: GDPR-Phase2-Java/src/main/java/com/gdpr/phase2/encoder/field/LangEncoder.java
package com.gdpr.phase2.encoder.field;


import com.gdpr.phase2.encoder.BaseEncoder;

public class LangEncoder implements BaseEncoder<String> {
    private LangEncoder() {
    }

    private IntEncoder intEncoder = IntEncoder.getInstance();
    private static final LangEncoder instance = new LangEncoder();
    public static LangEncoder getInstance() {
        return instance;
    }
//    public static String encode(String value,int numBits) {
//        value = value.toUpperCase();
//        final int ASCII_START = 65;
//        final int firstLetter = (int)value.charAt(0) - ASCII_START;
//        final int secondLetter = (int)value.charAt(1) - ASCII_START;
//        if( firstLetter < 0 || firstLetter > 25 || secondLetter < 0 || secondLetter > 25) {
//            // throw error
//        }
//        if(numBits % 2 == 1){
//            //throw error
//        }
//        numBits = numBits/2;
//        String firstLetterBString = IntEncoder.encode(firstLetter,numBits);
//        String secondLetterBString = IntEncoder.encode(secondLetter,numBits);
//        return firstLetterBString + secondLetterBString;
//    }

    public final String decode(String value) {
        String retr = "";

        if(value.length() % 2 == 0) {
            final int ASCII_START = 65;
            final int mid = value.length() / 2;
            final int firstLetter = intEncoder.decode(value.substring(0, mid)) + ASCII_START;
            final int secondLetter = intEncoder.decode(value.substring(mid)) + ASCII_START;
            retr = fromCharCode(firstLetter) + fromCharCode(secondLetter);
        } else {
            // throw error;
        }
        return retr;

    }

    private static String fromCharCode(int ...codePoints) {
        return new String(codePoints, 0, codePoints.length);
    }

}
